adrs/0049-prebuilt-rubies-at-runtime.md

```diff
@@ -0,0 +1,75 @@
+# ADR-0049 Prebuilt Rubies at Runtime
+
+***Status***: proposed
+
+## Context
+Currently, only two versions of Ruby are available at runtime as part of the hosted build VM tools cache.  The setup-ruby action only supports using what's in the cache and does not support pulling other versions at runtime.  It also doesn't support pulling other versions like JRuby at runtime.
+
+This is in contrast to other setup actions like setup-node where the cache is an optimization but supports pulling any version by a specified semver at runtime.
+
+## Goals
+
+- Offer as many versions as possible by platforms
+- Offer popular Ruby variants like JRuby
+- Consistent with other setup-xxx actions patterns and Actions workflows.
+
+## Decisions
+
+### Prebuilt Rubies
+
+Other tools like offer a [distribution](https://nodejs.org/dist/) which offers a [queryable endpoint](https://nodejs.org/dist/index.json) which allows the desired version [by a version spec which is semver](https://github.com/actions/setup-node#setup-node).
+
+Other CI services like Travis support pulling a wide variety of [ruby versions offered here](http://rubies.travis-ci.org/).
+
+Requirements:
+- Offer prebuilt Ruby, JRuby, TruffleRuby versions 
+- Setup-Ruby exposes `ruby-version` which accepts a semver
+- Rubies are discoverable and queryable by the Setup-Ruby action.
+- Initially offer for all platforms offered by the hosted Actions VMs (Ubuntu 18, Mac 10.15 and Windows 2019)
+
+NOTE: The prebuilt rubies offered will only be supported for use by GitHub Actions.
+
+The scripts and tooling to build the Rubies will be at `actions/build-ruby` and it will be consumed by the existing action `actions/setup-ruby`.  Since the capabilities are additive, there's no need to create a `v2`.  There is no compat break.
+
+A specific version of a Ruby will be offered as an individual semi-immutable build-ruby repo release similar to how the [actions runner exposes versions](https://github.com/actions/runner/releases/tag/v2.164.0) where each version offers n platforms.  This is also similar to [Travis individual versions](http://rubies.travis-ci.org/ubuntu/18.04/s390x/ruby-2.6.5).
+
+NOTE: semi-immutable means it's desirable to be immutable but it's possible to patch.  Note that 
+
+Option 1: GitHub GPR Universal Package
+
+The `actions/build-ruby` repo offers packages for [example](https://github.com/actions/setup-ruby/packages). 
+
+Option 2: GitHub Release
+
+The `actions/build-ruby` repo offers releases for [example](https://github.com/actions/runner/releases/tag/v2.164.0)
+
+NOTE: release assets are backed by a CDN
+
+Offering each version as an individual package / release offers queryable APIs and the ability to convey whether pre-release or not by version using the packaging and release features.
+
+This is fairly straight forward to come up with a scheme to complete automate with a workflow.
```

Comment:
It's not straightforward in the case building fails, as explained in https://github.com/actions/setup-ruby/issues/48#issuecomment-573427312.
How do you plan to address that? Modify the create-release and upload-release-asset to allow using them for patching an existing release?

Reply:
I believe the runner runs a graph of jobs (for each arch), each job uploads an artifact and then there's a job that depends on all those which finally aggregates all and creates the release.  If any job fails the job that creates the release doesn't run.

We could also easily do another approach which is to create the release in pre-release and then each job modifies the release and then either a final job node flips it or there's another canary set of workflows that validates all E2E functionality which flips from pre-release to release (runner does that).

Reply:
> each job uploads an artifact and then there's a job that depends on all those which finally aggregates all and creates the release.

Uploading release assets requires the release to be created already. It could be in draft mode, but there is no premade action under github/actions to turn a draft release into a public release AFAIK.
Using download-artifact/upload-artifact workarounds that at the cost of extra complexity.

I see, https://github.com/actions/runner/blob/a727194742dd7b28c265527f71e2a3669b71516a/.github/workflows/release.yml does something like you mention.
https://github.com/actions/github-script seems a way to workaround the limitations of the create-release and upload-release-asset actions, but it's more verbose and it's JavaScript in a YAML string. It also includes the list of platforms/versions twice.

Anyway, since I guess this is a part GitHub will maintain, best to leave it to you.
Readability would still be good if the community wants to help building new versions or understand how all this works.